### PR TITLE
refactor GNN to allow any output (inputs are always the same)

### PR DESCRIPTION
**Pull request by ronuchit**
should cause no functionality changes

one step toward #731

### CODE REVIEW

File: src/approaches/gnn_approach.py [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+                                   object_to_node: Dict) -> _Output:
+        """At evaluation time, given an output GNN, extract the actual
+        output."""
+        raise NotImplementedError("Override me!")
+
+    @property
+    def is_learning_based(self) -> bool:
+        return True
+
+    def learn_from_offline_dataset(self, dataset: Dataset) -> None:
```
Comment (tomsilver) [inline src/approaches/gnn_approach.py:115]: i'm assuming this function is basically the same as before

Comment (ronuchit) [reply src/approaches/gnn_approach.py:115]: yup

File: src/approaches/gnn_approach.py [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+        raise NotImplementedError("Override me!")
+
+    @abc.abstractmethod
+    def _graphify_single_target(self, datapoint: Tuple[State, Set[GroundAtom],
+                                                       Set[GroundAtom],
+                                                       _Output],
+                                graph_input: Dict,
+                                object_to_node: Dict) -> Dict:
+        """Given a single datapoint (input and output) and the return values of
+        graphify_single_input on the input, return a target graph."""
```
[the remark this answers is not in the document]
Comment (ronuchit) [reply src/approaches/gnn_approach.py:77]: i don't actually think it's necessary, removing

File: src/approaches/gnn_approach.py [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+        self._node_feature_to_index: Dict[Any, int] = {}
+        self._edge_feature_to_index: Dict[Any, int] = {}
+        self._input_normalizers: Dict = {}
+        self._target_normalizers: Dict = {}
+        self._data_exemplar: Tuple[Dict, Dict] = ({}, {})
+        # Seed torch.
+        torch.manual_seed(self._seed)
+
+    @abc.abstractmethod
+    def _extract_target_from_data(self, segment: Segment,
```
[the remark this answers is not in the document]
Comment (ronuchit) [reply src/approaches/gnn_approach.py:53]: will clarify the docstring to describe the relationship

**Pull request merged**